src/gpg_io.c

```diff
@@ -30,199 +27,196 @@
+        case IO_OFFSET_END:
+            G_gpg_vstate.io_offset = G_gpg_vstate.io_length;
+            break;
+        case IO_OFFSET_MARK:
+            G_gpg_vstate.io_offset = G_gpg_vstate.io_mark;
+            break;
+        default:
+            LEDGER_ASSERT(offset < G_gpg_vstate.io_length, "Bad offset!");
+            G_gpg_vstate.io_offset = offset;
+            break;
+    }
 }
 
 void gpg_io_mark() {
-  G_gpg_vstate.io_mark = G_gpg_vstate.io_offset;
+    G_gpg_vstate.io_mark = G_gpg_vstate.io_offset;
 }
 
 void gpg_io_inserted(unsigned int len) {
-  G_gpg_vstate.io_offset += len;
-  G_gpg_vstate.io_length += len;
+    G_gpg_vstate.io_offset += len;
+    G_gpg_vstate.io_length += len;
 }
 
 void gpg_io_discard(int clear) {
-  G_gpg_vstate.io_length = 0;
-  G_gpg_vstate.io_offset = 0;
-  G_gpg_vstate.io_mark   = 0;
-  if (clear) {
-    gpg_io_clear();
-  }
+    G_gpg_vstate.io_length = 0;
+    G_gpg_vstate.io_offset = 0;
+    G_gpg_vstate.io_mark = 0;
+    if (clear) {
+        gpg_io_clear();
+    }
 }
 
 void gpg_io_clear() {
-  os_memset(G_gpg_vstate.work.io_buffer, 0, GPG_IO_BUFFER_LENGTH);
+    memset(G_gpg_vstate.work.io_buffer, 0, GPG_IO_BUFFER_LENGTH);
 }
 
 /* ----------------------------------------------------------------------- */
 /* INSERT data to be sent                                                  */
 /* ----------------------------------------------------------------------- */
 
-void gpg_io_hole(unsigned int sz) {
-  if ((G_gpg_vstate.io_length + sz) > GPG_IO_BUFFER_LENGTH) {
-    THROW(ERROR_IO_FULL);
-    return;
-  }
-  os_memmove(G_gpg_vstate.work.io_buffer + G_gpg_vstate.io_offset + sz,
-             G_gpg_vstate.work.io_buffer + G_gpg_vstate.io_offset, G_gpg_vstate.io_length - G_gpg_vstate.io_offset);
-  G_gpg_vstate.io_length += sz;
+static void gpg_io_hole(unsigned int sz) {
+    LEDGER_ASSERT((G_gpg_vstate.io_length + sz) <= GPG_IO_BUFFER_LENGTH, "Bad hole!");
+    memmove(G_gpg_vstate.work.io_buffer + G_gpg_vstate.io_offset + sz,
+            G_gpg_vstate.work.io_buffer + G_gpg_vstate.io_offset,
+            G_gpg_vstate.io_length - G_gpg_vstate.io_offset);
+    G_gpg_vstate.io_length += sz;
 }
 
 void gpg_io_insert(unsigned char const *buff, unsigned int len) {
-  gpg_io_hole(len);
-  os_memmove(G_gpg_vstate.work.io_buffer + G_gpg_vstate.io_offset, buff, len);
-  G_gpg_vstate.io_offset += len;
+    gpg_io_hole(len);
+    memmove(G_gpg_vstate.work.io_buffer + G_gpg_vstate.io_offset, buff, len);
+    G_gpg_vstate.io_offset += len;
 }
 
 void gpg_io_insert_u32(unsigned int v32) {
-  gpg_io_hole(4);
-  G_gpg_vstate.work.io_buffer[G_gpg_vstate.io_offset + 0] = v32 >> 24;
-  G_gpg_vstate.work.io_buffer[G_gpg_vstate.io_offset + 1] = v32 >> 16;
-  G_gpg_vstate.work.io_buffer[G_gpg_vstate.io_offset + 2] = v32 >> 8;
-  G_gpg_vstate.work.io_buffer[G_gpg_vstate.io_offset + 3] = v32 >> 0;
-  G_gpg_vstate.io_offset += 4;
+    gpg_io_hole(4);
+    G_gpg_vstate.work.io_buffer[G_gpg_vstate.io_offset + 0] = v32 >> 24;
```

Comment:
Could be using U4BE_ENCODE from https://github.com/LedgerHQ/ledger-secure-sdk/blob/master/include/os_utils.h
Same for all insert and fetch methods.

Reply:
Fixed